File: leetcode/297.二叉树的序列化与反序列化/297.二叉树的序列化与反序列化.go
package main

import (
	"strconv"
	"strings"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type Codec struct {
}

func Constructor() (_ Codec) {
	return
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *TreeNode) string {
	// 以字符串方式储存
	sb := strings.Builder{}
	var serializePreOrder func(root *TreeNode)
	serializePreOrder = func(root *TreeNode) {
		// base case
		if root == nil {
			sb.WriteString("null,")
			return
		}
		// 前序遍历，转换int为字符串
		sb.WriteString(strconv.Itoa(root.Val))
		sb.WriteString(",")
		// 递归左右子树
		serializePreOrder(root.Left)
		serializePreOrder(root.Right)

	}
	serializePreOrder(root)
	return sb.String()
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *TreeNode {
	// 以“,”分割字符串，为字符串数组
	sp := strings.Split(data, ",")
	var deserializePreOrder func() *TreeNode
	deserializePreOrder = func() *TreeNode {
		// 前序遍历,并弹出index0
		if sp[0] == "null" {
			sp = sp[1:]
			return nil
		}
		// 转换str为int
		val, _ := strconv.Atoi(sp[0])
		sp = sp[1:]
		node := &TreeNode{Val: val}
		// 递归调用左右子树
		node.Left = deserializePreOrder()
		node.Right = deserializePreOrder()
		return node
	}
	return deserializePreOrder()
}

File: leetcode/297.二叉树的序列化与反序列化/297.二叉树的序列化与反序列化.py
from typing import List


class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Codec:
    def __init__(self):
        self.sep = ","
        self.null = "#"

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        res = []

        def serialize_preorder(root: TreeNode, res: List[str]):
            if root is None:
                res.append(self.null)
                return
            # 前序遍历位置
            res.append(str(root.val))
            serialize_preorder(root.left, res)
            serialize_preorder(root.right, res)

        serialize_preorder(root, res)
        return self.sep.join(res)

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        # 将字符串转化成列表
        nodes = data.split(self.sep)

        def deserialize_preorder(nodes: List[str]):
            if nodes == []:
                return None
            # 前序遍历位置
            # 列表最左侧就是根节点
            first = nodes.pop(0)
            if first == self.null:
                return None
            root = TreeNode(int(first))

            root.left = deserialize_preorder(nodes)
            root.right = deserialize_preorder(nodes)
            return root

        return deserialize_preorder(nodes)
